{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "scdefg.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "toc_visible": true,
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "accelerator": "GPU"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Munfred/worm-markers/blob/master/scdefg.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "GKkwa2wFWEiL"
      },
      "source": [
        "\n",
        "\n",
        "\n",
        "# scdefg: scvi on Colab for differential expression with a Flask GUI\n",
        "\n",
        "*This app uses [scvi-tools](https://scvi-tools.org) and is made to run on a\n",
        " Google Colab instance with a GPU.   \n",
        " The code for this app and more information is available at\n",
        "[https://github.com/munfred/scdefg](https://github.com/munfred/scdefg)*\n",
        "\n",
        "By using a GPU scvi-tools allows you to quickly perform differential expression \n",
        "on single cell RNA sequencing data.   \n",
        "Just select cell types and experiments to compare and some genes to highlight.   \n",
        "It will produce an interactive volcano plot and a csv file with results *in just a few seconds*! Try it! \n",
        "\n",
        "#### PS: Make sure to select GPU runtime!\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "iiyR_cbZnXWW",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "667d6cd8-f002-4f5d-e3e9-279cc88462be"
      },
      "source": [
        "# clone repo which contains the html templates for flask\n",
        "!git clone https://github.com/Munfred/scdefg.git"
      ],
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Cloning into 'scdefg'...\n",
            "remote: Enumerating objects: 78, done.\u001b[K\n",
            "remote: Counting objects: 100% (78/78), done.\u001b[K\n",
            "remote: Compressing objects: 100% (71/71), done.\u001b[K\n",
            "remote: Total 78 (delta 33), reused 10 (delta 4), pack-reused 0\u001b[K\n",
            "Unpacking objects: 100% (78/78), done.\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "1cljkrUaPeyg",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "e492a883-7089-4ed1-868a-2d49c35cdee8"
      },
      "source": [
        "# install flask dependencies on colab, takes a minute or two\n",
        "!pip install scvi-tools --quiet\n",
        "!pip install Flask-Misaka --quiet\n",
        "!pip install Flask-Markdown --quiet\n",
        "!pip install flask-ngrok --quiet\n",
        "!pip install flask==0.12.2 --quiet\n",
        "# Newer versions of flask don't work in Colab\n",
        "# See https://github.com/plotly/dash/issues/257"
      ],
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "\u001b[K     |████████████████████████████████| 153kB 8.9MB/s \n",
            "\u001b[K     |████████████████████████████████| 122kB 35.3MB/s \n",
            "\u001b[K     |████████████████████████████████| 194kB 29.4MB/s \n",
            "\u001b[K     |████████████████████████████████| 245kB 34.2MB/s \n",
            "\u001b[K     |████████████████████████████████| 51kB 7.9MB/s \n",
            "\u001b[K     |████████████████████████████████| 133kB 7.5MB/s \n",
            "\u001b[?25h  Building wheel for misaka (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "\u001b[K     |████████████████████████████████| 174kB 8.5MB/s \n",
            "\u001b[?25h  Building wheel for Flask-Markdown (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "\u001b[K     |████████████████████████████████| 92kB 6.1MB/s \n",
            "\u001b[?25h"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "v1ZYxoLvaxBb",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "011bb784-57fa-4c2e-f89e-9a5eb387aa9d"
      },
      "source": [
        "from flask import Flask, Response, jsonify, request, render_template, Blueprint, send_file, redirect, url_for\n",
        "import logging\n",
        "import pandas as pd\n",
        "import sys\n",
        "import json\n",
        "import time\n",
        "from io import StringIO\n",
        "import urllib\n",
        "from flaskext.markdown import Markdown\n",
        "from flask_misaka import Misaka\n",
        "from flask_ngrok import run_with_ngrok\n",
        "from google.colab import files\n",
        "import anndata\n",
        "import scvi\n",
        "import numpy as np\n",
        "from tqdm import tqdm\n",
        "import matplotlib.pyplot as plt\n",
        "import plotly.graph_objects as go\n",
        "import time\n",
        "print('scvi-tools verseion:', scvi.__version__)"
      ],
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "scvi-tools verseion: 0.8.1\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "xxiwiCB4GgQg",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "6bc19b93-b6b3-4ab8-e397-33bcf48ef7e0"
      },
      "source": [
        "# this cell will download the file taylor2020.h5ad with the C. elegans single cell \n",
        "# data from the CeNGEN project. The data is described in the preprint \n",
        "# \"Molecular topography of an entire nervous system\", by Taylor et al 2020\n",
        "# https://doi.org/10.1016/j.neuron.2018.07.042\n",
        "\n",
        "!wget -nc https://github.com/Munfred/wormcells-site/releases/download/taylor2020/taylor2020.h5ad\n",
        "\n",
        "# this is an example, you can just as easily load your own .h5ad file"
      ],
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2021-01-26 17:13:17--  https://github.com/Munfred/wormcells-site/releases/download/taylor2020/taylor2020.h5ad\n",
            "Resolving github.com (github.com)... 192.30.255.113\n",
            "Connecting to github.com (github.com)|192.30.255.113|:443... connected.\n",
            "HTTP request sent, awaiting response... 301 Moved Permanently\n",
            "Location: https://github.com/Munfred/wormcells-data/releases/download/taylor2020/taylor2020.h5ad [following]\n",
            "--2021-01-26 17:13:17--  https://github.com/Munfred/wormcells-data/releases/download/taylor2020/taylor2020.h5ad\n",
            "Reusing existing connection to github.com:443.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://github-production-release-asset-2e65be.s3.amazonaws.com/222185132/0b5c6880-5149-11eb-8695-20a4f0843b75?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20210126%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210126T171317Z&X-Amz-Expires=300&X-Amz-Signature=e2849b49a99e21edf9e320bab97db90719106ca5075ee2a2f34fe7c54de5730a&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=222185132&response-content-disposition=attachment%3B%20filename%3Dtaylor2020.h5ad&response-content-type=application%2Foctet-stream [following]\n",
            "--2021-01-26 17:13:17--  https://github-production-release-asset-2e65be.s3.amazonaws.com/222185132/0b5c6880-5149-11eb-8695-20a4f0843b75?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20210126%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210126T171317Z&X-Amz-Expires=300&X-Amz-Signature=e2849b49a99e21edf9e320bab97db90719106ca5075ee2a2f34fe7c54de5730a&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=222185132&response-content-disposition=attachment%3B%20filename%3Dtaylor2020.h5ad&response-content-type=application%2Foctet-stream\n",
            "Resolving github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)... 52.216.138.171\n",
            "Connecting to github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)|52.216.138.171|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 381516791 (364M) [application/octet-stream]\n",
            "Saving to: ‘taylor2020.h5ad’\n",
            "\n",
            "taylor2020.h5ad     100%[===================>] 363.84M  37.4MB/s    in 8.9s    \n",
            "\n",
            "2021-01-26 17:13:27 (40.8 MB/s) - ‘taylor2020.h5ad’ saved [381516791/381516791]\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ZED3NGP9GgNR",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 411
        },
        "outputId": "7f59a24a-0f63-4966-f5f7-a9f8a317e352"
      },
      "source": [
        "#here we load the adata which already has the two fields we will make our selection by:\n",
        "# `cell_type` and `experiment_code`\n",
        "# if you use your own adata, it MUST have these two columns named the same way\n",
        "# if they are named differently the code will break and you will need to change it by hand \n",
        "# in the rest of the notebook (it is not hard but might be annoying)\n",
        "adata = anndata.read('taylor2020.h5ad')\n",
        "print(adata)\n",
        "display(adata.obs.head())\n",
        "display(adata.var.head())"
      ],
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "AnnData object with n_obs × n_vars = 100955 × 46911\n",
            "    obs: 'dropbox_id', 'counts', 'experiment_code', 'cell_type', 'tissue'\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>dropbox_id</th>\n",
              "      <th>counts</th>\n",
              "      <th>experiment_code</th>\n",
              "      <th>cell_type</th>\n",
              "      <th>tissue</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>1806-ST-1-AAACCTGAGAGACGAA</th>\n",
              "      <td>1806-ST-1</td>\n",
              "      <td>65</td>\n",
              "      <td>Pan-1</td>\n",
              "      <td>Unannotated</td>\n",
              "      <td>Unannotated</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1806-ST-1-AAACCTGAGGTAAACT</th>\n",
              "      <td>1806-ST-1</td>\n",
              "      <td>367</td>\n",
              "      <td>Pan-1</td>\n",
              "      <td>AVF</td>\n",
              "      <td>Neuron</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1806-ST-1-AAACCTGAGGTAGCCA</th>\n",
              "      <td>1806-ST-1</td>\n",
              "      <td>1792</td>\n",
              "      <td>Pan-1</td>\n",
              "      <td>AVH</td>\n",
              "      <td>Neuron</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1806-ST-1-AAACCTGAGTAACCCT</th>\n",
              "      <td>1806-ST-1</td>\n",
              "      <td>1229</td>\n",
              "      <td>Pan-1</td>\n",
              "      <td>RIA</td>\n",
              "      <td>Neuron</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1806-ST-1-AAACCTGAGTACGCGA</th>\n",
              "      <td>1806-ST-1</td>\n",
              "      <td>1401</td>\n",
              "      <td>Pan-1</td>\n",
              "      <td>AUA</td>\n",
              "      <td>Neuron</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                           dropbox_id  counts  ...    cell_type       tissue\n",
              "1806-ST-1-AAACCTGAGAGACGAA  1806-ST-1      65  ...  Unannotated  Unannotated\n",
              "1806-ST-1-AAACCTGAGGTAAACT  1806-ST-1     367  ...          AVF       Neuron\n",
              "1806-ST-1-AAACCTGAGGTAGCCA  1806-ST-1    1792  ...          AVH       Neuron\n",
              "1806-ST-1-AAACCTGAGTAACCCT  1806-ST-1    1229  ...          RIA       Neuron\n",
              "1806-ST-1-AAACCTGAGTACGCGA  1806-ST-1    1401  ...          AUA       Neuron\n",
              "\n",
              "[5 rows x 5 columns]"
            ]
          },
          "metadata": {
            "tags": []
          }
        },
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>gene_id</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>WBGene00014450</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00014451</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00010957</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00010958</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00014452</th>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "Empty DataFrame\n",
              "Columns: []\n",
              "Index: [WBGene00014450, WBGene00014451, WBGene00010957, WBGene00010958, WBGene00014452]"
            ]
          },
          "metadata": {
            "tags": []
          }
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "-CrUMiF7LoHr",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 452
        },
        "outputId": "1c84803b-f70c-430b-b7c5-3d3f10e24aec"
      },
      "source": [
        "# our adata has annotations by experiment, by cell type and by tissue\n",
        "# we use the experiment code and cell type to generate a dataframe\n",
        "# that counts how many cells of each type are present in each experiment\n",
        "# this dataframe is needed to create the table that allows the user to select \n",
        "# which cells to compare on the app\n",
        "# cell type does on the index, experiment on the columns, entries are number of\n",
        "# cells of that type per experiment\n",
        "\n",
        "experiments= np.sort(adata.obs.experiment_code.unique())\n",
        "unique_cell_types=np.sort(adata.obs.cell_type.unique())\n",
        "census=pd.DataFrame(np.nan, index= unique_cell_types, columns=experiments)\n",
        "for experiment in experiments:\n",
        "    census[experiment] = census.index.map(adata.obs[adata.obs['experiment_code']==experiment]['cell_type'].value_counts())\n",
        "census.index=census.index.rename('Cell Type')\n",
        "census.to_csv('census.csv')\n",
        "census"
      ],
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Pan-1</th>\n",
              "      <th>Pan-2</th>\n",
              "      <th>acr-2</th>\n",
              "      <th>ceh-28_dat-1</th>\n",
              "      <th>ceh-34</th>\n",
              "      <th>cho-1_1</th>\n",
              "      <th>cho-1_2</th>\n",
              "      <th>eat-4</th>\n",
              "      <th>ift-20</th>\n",
              "      <th>nlp-13_ceh-2</th>\n",
              "      <th>nmr-1</th>\n",
              "      <th>tph-1_ceh-10</th>\n",
              "      <th>unc-3</th>\n",
              "      <th>unc-47_1</th>\n",
              "      <th>unc-47_2</th>\n",
              "      <th>unc-53</th>\n",
              "      <th>unc-86</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Cell Type</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>ADA</th>\n",
              "      <td>91</td>\n",
              "      <td>130</td>\n",
              "      <td>10</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "      <td>2</td>\n",
              "      <td>4</td>\n",
              "      <td>115</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>4</td>\n",
              "      <td>1</td>\n",
              "      <td>63</td>\n",
              "      <td>0</td>\n",
              "      <td>5</td>\n",
              "      <td>3</td>\n",
              "      <td>174</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ADE</th>\n",
              "      <td>48</td>\n",
              "      <td>70</td>\n",
              "      <td>0</td>\n",
              "      <td>282</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>50</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ADF</th>\n",
              "      <td>3</td>\n",
              "      <td>16</td>\n",
              "      <td>12</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "      <td>3</td>\n",
              "      <td>3</td>\n",
              "      <td>7</td>\n",
              "      <td>243</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>464</td>\n",
              "      <td>3</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ADL</th>\n",
              "      <td>8</td>\n",
              "      <td>14</td>\n",
              "      <td>11</td>\n",
              "      <td>5</td>\n",
              "      <td>7</td>\n",
              "      <td>6</td>\n",
              "      <td>7</td>\n",
              "      <td>330</td>\n",
              "      <td>512</td>\n",
              "      <td>3</td>\n",
              "      <td>2</td>\n",
              "      <td>1</td>\n",
              "      <td>28</td>\n",
              "      <td>0</td>\n",
              "      <td>5</td>\n",
              "      <td>12</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AFD</th>\n",
              "      <td>27</td>\n",
              "      <td>57</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>141</td>\n",
              "      <td>3</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>3</td>\n",
              "      <td>0</td>\n",
              "      <td>3</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>VD_DD</th>\n",
              "      <td>125</td>\n",
              "      <td>153</td>\n",
              "      <td>39</td>\n",
              "      <td>30</td>\n",
              "      <td>6</td>\n",
              "      <td>5</td>\n",
              "      <td>8</td>\n",
              "      <td>11</td>\n",
              "      <td>3</td>\n",
              "      <td>3</td>\n",
              "      <td>5</td>\n",
              "      <td>5</td>\n",
              "      <td>31</td>\n",
              "      <td>594</td>\n",
              "      <td>464</td>\n",
              "      <td>31</td>\n",
              "      <td>12</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Vulval_cells</th>\n",
              "      <td>3</td>\n",
              "      <td>3</td>\n",
              "      <td>1</td>\n",
              "      <td>13</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>56</td>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Vulval_muscle</th>\n",
              "      <td>2</td>\n",
              "      <td>3</td>\n",
              "      <td>5</td>\n",
              "      <td>11</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "      <td>3</td>\n",
              "      <td>2</td>\n",
              "      <td>20</td>\n",
              "      <td>2</td>\n",
              "      <td>9</td>\n",
              "      <td>5</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>42</td>\n",
              "      <td>9</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>XXX</th>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "      <td>12</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>2</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>hmc</th>\n",
              "      <td>2</td>\n",
              "      <td>1</td>\n",
              "      <td>4</td>\n",
              "      <td>10</td>\n",
              "      <td>2</td>\n",
              "      <td>4</td>\n",
              "      <td>1</td>\n",
              "      <td>3</td>\n",
              "      <td>0</td>\n",
              "      <td>10</td>\n",
              "      <td>0</td>\n",
              "      <td>3</td>\n",
              "      <td>9</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>169 rows × 17 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "               Pan-1  Pan-2  acr-2  ...  unc-47_2  unc-53  unc-86\n",
              "Cell Type                           ...                          \n",
              "ADA               91    130     10  ...         5       3     174\n",
              "ADE               48     70      0  ...         0       2       0\n",
              "ADF                3     16     12  ...         0       1       1\n",
              "ADL                8     14     11  ...         5      12       4\n",
              "AFD               27     57      2  ...         3       2       0\n",
              "...              ...    ...    ...  ...       ...     ...     ...\n",
              "VD_DD            125    153     39  ...       464      31      12\n",
              "Vulval_cells       3      3      1  ...         0       0      10\n",
              "Vulval_muscle      2      3      5  ...         1      42       9\n",
              "XXX                1      1      2  ...         1       0       1\n",
              "hmc                2      1      4  ...         0       1       1\n",
              "\n",
              "[169 rows x 17 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 6
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "4P4EEdYiGgDF",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 398
        },
        "outputId": "c540e101-afcc-4a7e-eb5b-2a91dceb3a5c"
      },
      "source": [
        "# in order to create the interactive volcano plot with gene names and descriptions\n",
        "# we download the C. elegans gene descriptions from the Alliance of Genome Resources\n",
        "!wget -nc http://download.alliancegenome.org/3.2.0/GENE-DESCRIPTION-TSV/WB/GENE-DESCRIPTION-TSV_WB_10.tsv.gz\n",
        "# then we parse the gene descriptions into a dataframe and add those to the adata\n",
        "# the alliance download page at https://www.alliancegenome.org/downloads  provides\n",
        "# gene descriptions for C.elegans, zebrafish, drosophilam humans, rat, mouse and yeast\n",
        "# so by tweaking the code in this cell you can add gene descriptions to your own dataset!\n",
        "gene_descriptions=pd.read_csv('GENE-DESCRIPTION-TSV_WB_10.tsv.gz', sep='\\t', skiprows=14, header=None)\n",
        "gene_descriptions.columns=['orig_id','gene_name', 'gene_description']\n",
        "gene_descriptions['gene_id']=gene_descriptions['orig_id'].str.split(':',expand=True)[1]\n",
        "gene_descriptions.index=gene_descriptions['gene_id'].values\n",
        "\n",
        "# we add the gene descriptions to adata.var by mapping them using the gene id that\n",
        "# is already in the adata.var index\n",
        "adata.var['gene_id']=adata.var.index.map(gene_descriptions['gene_id'])\n",
        "adata.var['gene_name']=adata.var.index.map(gene_descriptions['gene_name'])\n",
        "adata.var['gene_description']=adata.var.index.map(gene_descriptions['gene_description'])\n",
        "adata.var.head()"
      ],
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2021-01-26 17:13:27--  http://download.alliancegenome.org/3.2.0/GENE-DESCRIPTION-TSV/WB/GENE-DESCRIPTION-TSV_WB_10.tsv.gz\n",
            "Resolving download.alliancegenome.org (download.alliancegenome.org)... 13.224.2.23, 13.224.2.7, 13.224.2.26, ...\n",
            "Connecting to download.alliancegenome.org (download.alliancegenome.org)|13.224.2.23|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 885437 (865K) [application/x-gzip]\n",
            "Saving to: ‘GENE-DESCRIPTION-TSV_WB_10.tsv.gz’\n",
            "\n",
            "GENE-DESCRIPTION-TS 100%[===================>] 864.68K  2.59MB/s    in 0.3s    \n",
            "\n",
            "2021-01-26 17:13:28 (2.59 MB/s) - ‘GENE-DESCRIPTION-TSV_WB_10.tsv.gz’ saved [885437/885437]\n",
            "\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>gene_id</th>\n",
              "      <th>gene_name</th>\n",
              "      <th>gene_description</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>gene_id</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>WBGene00014450</th>\n",
              "      <td>WBGene00014450</td>\n",
              "      <td>MTCE.1</td>\n",
              "      <td>No description available</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00014451</th>\n",
              "      <td>WBGene00014451</td>\n",
              "      <td>MTCE.2</td>\n",
              "      <td>No description available</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00010957</th>\n",
              "      <td>WBGene00010957</td>\n",
              "      <td>nduo-6</td>\n",
              "      <td>Predicted to have NADH dehydrogenase (ubiquino...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00010958</th>\n",
              "      <td>WBGene00010958</td>\n",
              "      <td>ndfl-4</td>\n",
              "      <td>Predicted to have NADH dehydrogenase (ubiquino...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00014452</th>\n",
              "      <td>WBGene00014452</td>\n",
              "      <td>MTCE.5</td>\n",
              "      <td>No description available</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                       gene_id  ...                                   gene_description\n",
              "gene_id                         ...                                                   \n",
              "WBGene00014450  WBGene00014450  ...                           No description available\n",
              "WBGene00014451  WBGene00014451  ...                           No description available\n",
              "WBGene00010957  WBGene00010957  ...  Predicted to have NADH dehydrogenase (ubiquino...\n",
              "WBGene00010958  WBGene00010958  ...  Predicted to have NADH dehydrogenase (ubiquino...\n",
              "WBGene00014452  WBGene00014452  ...                           No description available\n",
              "\n",
              "[5 rows x 3 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 7
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "hs4nVRkEkRcC"
      },
      "source": [
        "# Train the scvi model \n",
        "Here we have 100k cells and ~11k genes, so 5 epochs with scvi is sufficient\n",
        "- Make sure to be running the notebook with GPU runtime, the model will train in a few minutes\n",
        "- If using your own data and you want to account for batches, make sure that `batch_key` matches the experiment name from the adata\n",
        "- If your data has less cells, you will need more epochs. Check the train/test error plot to ensure the model finished training"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 703
        },
        "id": "h6erYQZCjbwE",
        "outputId": "9ef91d28-244e-4877-aed9-2a2fa9d5a7a0"
      },
      "source": [
        "# filter genes with less than 500 counts across all experiments\n",
        "min_gene_counts=500\n",
        "print(adata)\n",
        "print('REMOVING GENES WITH LESS THAN ', min_gene_counts, ' COUNTS')\n",
        "adata = adata[:, adata.X.sum(0) > min_gene_counts].copy()\n",
        "print(adata)\n",
        "\n",
        "### PREPARE DATA\n",
        "adata.layers[\"counts\"] = adata.X.copy() # preserve counts\n",
        "scvi.data.setup_anndata(adata, layer=\"counts\", batch_key='experiment_code') \n",
        "\n",
        "### DEFINE MODEL\n",
        "model = scvi.model.SCVI(adata,  \n",
        "                        n_hidden=128, \n",
        "                        n_layers=2,\n",
        "                        gene_likelihood='nb',\n",
        "                        dispersion='gene-batch'\n",
        "                        )\n",
        "\n",
        "# MODEL TRAINING\n",
        "model.train(frequency=1, \n",
        "            n_epochs = 5,\n",
        "            lr=2e-3,\n",
        "            n_epochs_kl_warmup=None)\n",
        "\n",
        "train_test_results = pd.DataFrame(model.trainer.history).rename(columns={'elbo_train_set':'Train', 'elbo_test_set':'Test'})\n",
        "print(train_test_results)\n",
        "train_test_results.to_csv('train_test_results.csv')\n",
        "train_test_results.plot(logy=True)\n",
        "\n",
        "### SAVE AND LOAD MODELS\n",
        "# uncomment the lines below to save or load a trained model\n",
        "\n",
        "# model_name = \"my_model\"\n",
        "# model.save(model_name)\n",
        "# model = scvi.model.SCVI.load(adata, model_name, use_cuda=True)\n"
      ],
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "AnnData object with n_obs × n_vars = 100955 × 46911\n",
            "    obs: 'dropbox_id', 'counts', 'experiment_code', 'cell_type', 'tissue'\n",
            "    var: 'gene_id', 'gene_name', 'gene_description'\n",
            "REMOVING GENES WITH LESS THAN  500  COUNTS\n",
            "AnnData object with n_obs × n_vars = 100955 × 11569\n",
            "    obs: 'dropbox_id', 'counts', 'experiment_code', 'cell_type', 'tissue'\n",
            "    var: 'gene_id', 'gene_name', 'gene_description'\n",
            "\u001b[34mINFO    \u001b[0m Using batches from adata.obs\u001b[1m[\u001b[0m\u001b[32m\"experiment_code\"\u001b[0m\u001b[1m]\u001b[0m                        \n",
            "\u001b[34mINFO    \u001b[0m No label_key inputted, assuming all cells have same label              \n",
            "\u001b[34mINFO    \u001b[0m Using data from adata.layers\u001b[1m[\u001b[0m\u001b[32m\"counts\"\u001b[0m\u001b[1m]\u001b[0m                                 \n",
            "\u001b[34mINFO    \u001b[0m Computing library size prior per batch                                 \n",
            "\u001b[34mINFO    \u001b[0m Successfully registered anndata object containing \u001b[1;34m100955\u001b[0m cells, \u001b[1;34m11569\u001b[0m  \n",
            "         vars, \u001b[1;34m17\u001b[0m batches, \u001b[1;34m1\u001b[0m labels, and \u001b[1;34m0\u001b[0m proteins. Also registered \u001b[1;34m0\u001b[0m extra    \n",
            "         categorical covariates and \u001b[1;34m0\u001b[0m extra continuous covariates.              \n",
            "\u001b[34mINFO    \u001b[0m Please do not further modify adata until model is trained.             \n",
            "\u001b[34mINFO    \u001b[0m Training for \u001b[1;34m5\u001b[0m epochs                                                  \n",
            "\u001b[34mINFO    \u001b[0m Training without KL warmup                                             \n",
            "Training...: 100%|██████████| 5/5 [01:13<00:00, 14.66s/it]\n",
            "\u001b[34mINFO    \u001b[0m Training time:  \u001b[1;34m45\u001b[0m s. \u001b[35m/\u001b[0m \u001b[1;34m5\u001b[0m epochs                                       \n",
            "         Train         Test\n",
            "0  4126.462597  4067.239430\n",
            "1  1700.068881  1666.080242\n",
            "2  1629.731851  1602.512449\n",
            "3  1594.055509  1579.411283\n",
            "4  1577.971812  1566.330906\n",
            "5  1574.795757  1557.486543\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<matplotlib.axes._subplots.AxesSubplot at 0x7f3a76344978>"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 8
        },
        {
          "output_type": "display_data",
          "data": {
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYcAAAD4CAYAAAAHHSreAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAgAElEQVR4nO3de5TU9X3/8ed7rntld9kFRRZZFEQYNWAwVo0RNSjekElCqj9M2miTmtaaps3PmqY5ac9pz0l72qY1zS+Jbf3l9NeemDSpiYLRGqP1gmG5yB1BlEUWloVd2IW978x8fn/MAHuFvc1+5/J6nLNn5vudz/f7fY8Cr/28v9/5jjnnEBER6cvndQEiIpJ5FA4iIjKIwkFERAZROIiIyCAKBxERGSTgdQETpaqqytXU1HhdhohIVtm0aVOTc27awPU5Ew41NTVs3LjR6zJERLKKmR0Yar3aSiIiMojCQUREBlE4iIjIIDlzzkFEZLR6e3upr6+nq6vL61LSrqCggOrqaoLB4IjGKxxEJG/V19dTWlpKTU0NZuZ1OWnjnKO5uZn6+nrmzJkzom3UVhKRvNXV1UVlZWVOBwOAmVFZWTmqGZLCQUTyWq4Hw2mjfZ95Hw7r9jXxxMvvel2GiEhGyftwWL//ON/65V6Onsr9E1Iiklmam5tZtGgRixYt4sILL2TmzJlnlnt6es657caNG3n00UfTVlven5C++6oZ/OPL7/KL7Uf4retrvC5HRPJIZWUlW7ZsAeDP//zPKSkp4Stf+cqZ12OxGIHA0P9ML1myhCVLlqSttryfOcxreYO/m/Jj1mw77HUpIiL89m//Ng8//DDXXnstjz32GLW1tVx33XUsXryY66+/nj179gDw6quvcvfddwPJYHnwwQdZunQpl1xyCU888cS468jomYOZLQC+BFQBLzvnvjvhB2ncySd7fsbf1t3KkdarubCsYMIPISKZ7y+e28muwycndJ8LL5rCN+6JjHq7+vp61q1bh9/v5+TJk7z++usEAgF++ctf8qd/+qf89Kc/HbTNO++8wyuvvMKpU6eYP38+X/ziF0f8mYahjHjmYGZ+M3vbzNaM9WBm9pSZHTWzHUO8ttzM9pjZPjN7HMA5t9s59zDwaeCGsR73nCJRAO70r2ft9oa0HEJEZDRWrVqF3+8HoLW1lVWrVnHFFVfw5S9/mZ07dw65zV133UU4HKaqqorp06fT2Ng4rhpGM3P4ErAbmDLwBTObDnQ65071WTfXObdvwNAfAP8E/NuA7f3Ad4BlQD2wwcyedc7tMrMVwBeB/zeKWkeu8lK48Co+dWwDX9t2mIc+OrIPiIhIbhnLb/jpUlxcfOb517/+dW6++WaeeeYZ6urqWLp06ZDbhMPhM8/9fj+xWGxcNYxo5mBm1cBdwL8MM+Qm4GdmFk6N/zzw7YGDnHOvAceH2P4jwD7n3PvOuR7gaeDe1DbPOufuAFaPpNYxiURZEN9D4wfvcqilM22HEREZrdbWVmbOnAnAD37wg0k77kjbSv8APAYkhnrROfefwIvAj8xsNfAgsGoUdcwEDvZZrgdmmtlSM3vCzL4PPD/UhmZ2j5k92draOorDDRBZCcAd/lrW6sS0iGSQxx57jK9+9assXrx43LOB0TDn3LkHmN0N3Omc+z0zWwp8xTl39zBjnwbuBC51zh0bZkwNsMY5d0WfdZ8Cljvnfie1/BngWufcIyN9I0uWLHHj+rKf79/EnmMdPFb+LX7+yEfHvh8RyRq7d+9mwYIFXpcxaYZ6v2a2yTk36JrYkcwcbgBWmFkdyXbPLWb27wMHmdmNwBXAM8A3RlnzIWBWn+Xq1LrJE4kyP7aX5kPv8kFzx6QeWkQk05w3HJxzX3XOVTvnaoD7gF855x7oO8bMFgNPkjxP8Dmg0sz+chR1bADmmdkcMwuljvPsKLYfv9RVS3f51rNmu1pLIpLfJupDcEXAp51z7znnEsBngUHfS2pmPwTeAuabWb2ZPQTgnIsBj5A8b7Eb+LFzbujrtdKlYjbM/DCrCjewdpsuaRWR/DaqD8E5514FXh1i/ZsDlnuBfx5i3P3n2PfzDHPSedJEosw99GecaniX/U1XM6eq+PzbiIjkoLy/fUY/C+8Fkq0lXbUkIvlM4dBX+cVQfQ2rCjewRq0lEcljCoeBIlEuib1HV+Ne9h09df7xIiJjNJ5bdkPy5nvr1q1LS20Kh4FOt5b86zV7EJG0On3L7i1btvDwww/z5S9/+cxyKBQ67/YKh8lUVg2zrmVVwUbWbGvgfB8SFBGZSJs2beKmm27iwx/+MLfffjsNDclfUp944gkWLlzIVVddxX333UddXR3f+973+Na3vsWiRYt4/fXXJ7SOjL5lt2ciUWoOPk7i2F72NC7m8gsH3WtQRHLNLx6HI9sndp8XXgl3fHPEw51z/MEf/AE///nPmTZtGj/60Y/42te+xlNPPcU3v/lN9u/fTzgcpqWlhfLych5++OFBXxA0URQOQ1l4L7zwOHf517N228cUDiIyKbq7u9mxYwfLli0DIB6PM2PGDACuuuoqVq9ezcqVK1m5cmXaa1E4DGXKRXDxdaxq2MhvbWvgj5Zdhpl5XZWIpNMofsNPF+cckUiEt956a9Bra9eu5bXXXuO5557jr/7qr9i+fYJnOQPonMNwIlEu7t2Pr3kvuxom9tuhRESGEg6HOXbs2Jlw6O3tZefOnSQSCQ4ePMjNN9/MX//1X9Pa2kpbWxulpaWcOpWeqyoVDsNZsAKHcbeuWhKRSeLz+fjJT37Cn/zJn/ChD32IRYsWsW7dOuLxOA888ABXXnklixcv5tFHH6W8vJx77rmHZ555Ji0npM97y+5sMe5bdg/l/95J/eFD3B/8Fq/975vVWhLJMbpl9/hu2Z2/IlGqe+soOPEu2w+N48uERESyjMLhXBaswJmPFYFfq7UkInlF4XAupRdgs2/gkwUbWbv1sD4QJ5KD8uXv9Wjfp8LhfCJRLur9gJKTe3n7YIvX1YjIBCooKKC5uTnnA8I5R3NzMwUFBSPeRp9zOJ8FK3DPf4UVwfWs2XoTV19c4XVFIjJBqqurqa+v59ixIb/yPqcUFBRQXV094vEKh/MpmYbV3Mgn6zewctth/uyuBfh8umpJJBcEg0HmzJnjdRkZSW2lkYhEubC3noq2vWz64ITX1YiIpJ3CYSQW3IMzf6q1pG+IE5Hcp3AYieIqbM7H+GR4A89vbyCeyO2TVyIiCoeRikSZ3nuI6e17qd1/3OtqRETSSuEwUqnW0r3B9azdrtaSiOQ2hcNIFU3FLllKNLyBX2xrIBZPeF2RiEjaKBxGIxJlWu9hZnTu4dfvq7UkIrlL4TAal9+F8wWIhmrVWhKRnKZwGI2iqdglN7MyVMsvtjfQq9aSiOQohcNoRaJU9h7h4q49vLmvyetqRETSQuEwWpffhfMFiYZrWavbeItIjlI4jFZhOTb3Vu4N1vLizgZ6YmotiUjuUTiMRSTK1N5GLu1+hzf25f7dHEUk/ygcxmL+HTh/iE+Ea1mzVa0lEck9CoexKCjD5n6cewK1vLSrga7euNcViYhMKIXDWEWilMeOMa/nHV7bq9aSiOQWhcNYXbYc5w/zyXAta3TVkojkGIXDWBVMweYt4+5ALS/vbqCzR60lEckdCofxiEQpizWxoHc3r+456nU1IiITRuEwHpfdjgsU8KmCDWotiUhOUTiMR7gUm7eMO/21vPJOAx09Ma8rEhGZEAqH8YpEmRJr5orYbl7erdaSiOQGhcN4zbsdFyjkU4UbWLNNt/EWkdygcBivcAl22W3c6avlf/Y00tat1pKIZD+Fw0SIRCmJHWdRYhe/3NXodTUiIuOmcJgI827HBYtYVaDWkojkBoXDRAgVYZct53ZfLW/ubaS1s9frikRExkXhMFEiUUpiJ1jsdvKSWksikuUUDhNl3jJcsJjfLNzIWrWWRCTLKRwmSrAQm38Hy2w9695tpKWjx+uKRETGTOEwkSJRimKtXMNOXtx5xOtqRETGTOEwkeZ+HBcq4b6ijbrXkohktYwOBzNbYGbfM7OfmNkXva7nvIIF2Pw7uZX11L7XSHNbt9cViYiMyXnDwcwKzKzWzLaa2U4z+4uxHszMnjKzo2a2Y4jXlpvZHjPbZ2aPAzjndjvnHgY+Ddww1uNOqkiUwthJrmUHL+7UVUsikp1GMnPoBm5xzn0IWAQsN7Pf6DvAzKabWemAdXOH2NcPgOUDV5qZH/gOcAewELjfzBamXlsBrAWeH0Gt3rv0Flx4CvcVb9IH4kQka503HFxSW2oxmPpxA4bdBPzMzMIAZvZ54NtD7Os14PgQh/kIsM85975zrgd4Grg3tc2zzrk7gNVD1Wdm95jZk62tred7K5Mj1Vq6xdWy6f1Gjp1Sa0lEss+IzjmYmd/MtgBHgZecc+v7vu6c+0/gReBHZrYaeBBYNYo6ZgIH+yzXAzPNbKmZPWFm32eYmYNz7jnn3BfKyspGcbg0i0QpiJ3kOtvBCzt0YlpEss+IwsE5F3fOLQKqgY+Y2RVDjPkboAv4LrCiz2xjzJxzrzrnHnXO/a5z7jvj3d+kufRmCJdxf/EmntNVSyKShUZ1tZJzrgV4haHPG9wIXAE8A3xjlHUcAmb1Wa5OrctOgTBcfhdLE+vZUneUxpNdXlckIjIqI7laaZqZlaeeFwLLgHcGjFkMPEnyPMHngEoz+8tR1LEBmGdmc8wsBNwHPDuK7TNPJEo43sYNtp3nt2v2ICLZZSQzhxnAK2a2jeQ/4i8559YMGFMEfNo5955zLgF8FjgwcEdm9kPgLWC+mdWb2UMAzrkY8AjJ8xa7gR8753aO9U1lhEuWQkE5q0s26QNxIpJ1Aucb4JzbBiw+z5g3Byz3Av88xLj7z7GP58mWy1VHIhCCBXdz47Zn+L0DRznc0slF5YVeVyUiMiIZ/QnprBeJEo63c6Nvm1pLIpJVFA7pNOcmKKzggRJdtSQi2UXhkE7+ICy4hxvitbxz8CgHj3d4XZGIyIgoHNItEiUU7+Am31bWqrUkIllC4ZBuNR+DwqmsLt2sey2JSNZQOKSbPwALV3B9rJZ3DzVR19TudUUiIuelcJgMkSjBeCdLfVvUWhKRrKBwmAyzPwpFVXymdLM+ECciWUHhMBlSraVre2vZ33CM946N+56EIiJppXCYLJEowUQXt/i3sGarZg8iktkUDpNl9g1QPI3PlG5m7XZdtSQimU3hMFl8flh4L9f0bOBgYxN7G095XZGIyLAUDpMpEiWQ6OJW/xbWbNXsQUQyl8JhMl18HZRcmLxqaXsDzg38Km4RkcygcJhMqdbSh3s2cORYM7sb1FoSkcykcJhskSiBRDfLAm/rxLSIZCyFw2SbdS2UzuCB1Afi1FoSkUykcJhsPh8sXMnV3Rtobm5ix6GTXlckIjKIwsELkSj+RA+3Bd5mjVpLIpKBFA5eqL4GpsxMfiBOrSURyUAKBy+kWktXdW+k9UQzW+tbva5IRKQfhYNXIlH8iV6WBzbrA3EiknEUDl6pXgJls3igdDNrtzeQSKi1JCKZQ+HgFTNYeC9Xdm2kvbWZtw+e8LoiEZEzFA5einwCn4txR3Azz+k23iKSQRQOXpp5NZRdzAOlm3l+ewNxtZZEJEMoHLxkBpGVRDo30XWqmY11x72uSEQEUDh4LxLF52LcHdL3S4tI5lA4eO2ixVBRw+qSzfxih1pLIpIZFA5eM4NIlAWdm4m1NbP+/WavKxIRUThkhD6tpefUWhKRDKBwyAQXXgVTL2F1yWZe2NFALJ7wuiIRyXMKh0yQai3N79yMdTSz7j21lkTEWwqHTBGJ4nNx7g1vZs023WtJRLylcMgUF1wBlXO5v3gTL+5spCem1pKIeEfhkClSraV5nW8T6GzizX1NXlckInlM4ZBJIlHMJVhZsEkfiBMRTykcMsn0hVB1GfcXb+K/dx2hOxb3uiIRyVMKh0ySai1d2rGVgq4mXtur1pKIeEPhkGlSraVo4WbW6qolEfGIwiHTTF8A0y7nvqKNvLSrka5etZZEZPIpHDJRJMqc9q0U9TTx6p6jXlcjInlI4ZCJIlEMx6cKdRtvEfGGwiETTZsP0yP8ZtFGXt59lI6emNcViUieUThkqkiU2e3bmNJ7jFfeOeZ1NSKSZxQOmSqyEsOxqkj3WhKRyadwyFRV8+CCK1lVuJFfvXOUtm61lkRk8igcMllkJbPbt1ERO8bLuxu9rkZE8ojCIZNFogDcV6yrlkRkcikcMlnlpXDhVXyqYAP/s+cYJ7t6va5IRPKEwiHTRaJUt+9gWryRX+5Sa0lEJofCIdNFVgJwX8nbai2JyKRROGS6qZfAjEV8IlzL6+8eo7VDrSURST+FQzaIRJnZvosLEkd5cdcRr6sRkTygcMgGqdbS/yrRVUsiMjkUDtmgogYuuppoqJY39zVxor3H64pEJMcpHLJFJMqM9t1c5I7wwk61lkQkvRQO2SLVWnqg9G3WqrUkImmmcMgW5RdD9TWsDNay7r0mmtq6va5IRHKYwiGbRKJc0P4OszjCL3aotSQi6aNwyCYL7wXgs1PeZq1u4y0iaaRwyCZl1TDrWu4JrGf9/uMcPdnldUUikqMUDtkmEmV6+15qaOD57ToxLSLpoXDINqnW0m+Vvc1ahYOIpInCIdtMuQguvo67/b9mQ90JGlo7va5IRHKQwiEbRaJUte/jUjvE89t11ZKITDyFQzZasAIwPle+hTW6aklE0kDhkI2mzIDZ13On/Zq3P2ih/kSH1xWJSI5ROGSrSJSpHe8xz+p11ZKITDiFQ7ZKtZYeLN+i23iLyIRTOGSr0gug5qMst7fYVt/CgeZ2rysSkRyicMhmkZVUdOznMqvXZx5EZEJldDiY2QIz+56Z/cTMvuh1PRlnwQowH79TsYU1WxUOIjJxzhsOZjbLzF4xs11mttPMvjTWg5nZU2Z21Mx2DPHacjPbY2b7zOxxAOfcbufcw8CngRvGetycVTIdam7kNt5iV0Mr7x9r87oiEckRI5k5xIA/ds4tBH4D+H0zW9h3gJlNN7PSAevmDrGvHwDLB640Mz/wHeAOYCFw/+ljmNkKYC3w/AhqzT+RKOUddVxuB/UlQCIyYc4bDs65Bufc5tTzU8BuYOaAYTcBPzOzMICZfR749hD7eg04PsRhPgLsc86975zrAZ4G7k1t86xz7g5g9VD1mdk9ZvZka2vr+d5KblpwD5ifz0/VVUsiMnFGdc7BzGqAxcD6vuudc/8JvAj8yMxWAw8Cq0ax65nAwT7L9cBMM1tqZk+Y2fcZZubgnHvOOfeFsrKyURwuhxRXwZyP8XG3jj2NJ3m38ZTXFYlIDhhxOJhZCfBT4A+dcycHvu6c+xugC/gusMI5N+4GuHPuVefco86533XOfWe8+8tZkShlHR+w0HdAswcRmRAjCgczC5IMhv9wzv3XMGNuBK4AngG+Mco6DgGz+ixXp9bJSKRaS1+YupU12w7jnPO6IhHJciO5WsmAfwV2O+f+fpgxi4EnSZ4n+BxQaWZ/OYo6NgDzzGyOmYWA+4BnR7F9fiuaCpcs5dbEOt471sYetZZEZJxGMnO4AfgMcIuZbUn93DlgTBHwaefce865BPBZ4MDAHZnZD4G3gPlmVm9mDwE452LAIyTPW+wGfuyc2znmd5WPIlFKOw5ylW+/PvMgIuNmudKCWLJkidu4caPXZXin4zj87TyeK/oEf89qfvXHN5Gc9ImIDM/MNjnnlgxcn9GfkJZRKJoKl9zMzfE32d/Uxs7Dg64ZEBEZMYVDLolEKek8xGL/fl21JCLjonDIJZffCb4gn5+6lbXbddWSiIydwiGXFFbApbdwU+wNDh7vYFt9nn5qXETGTeGQayJRijsbWBJ4X7fxFpExUzjkmsvvBH8o2Vra1qDWkoiMicIh1xSUwdyPc2PvGxxuaWfzBy1eVyQiWUjhkIsiUYo6j3BN4H3dxltExkThkIsuWw7+MF+YupXntzeQSKi1JCKjo3DIRQVTYN4ybuh5g8aTHWw8cMLrikQkyygcclUkSmFXI78RfI+12w57XY2IZBmFQ6667HYIFPCFyq08v+MIcbWWRGQUFA65KlwK85ZxXdfrNJ3qpHb/UN/OKiIyNIVDLotEKeg6xkdD+1ij1pKIjILCIZfNux0ChXx+6hZe2HGEWDzhdUUikiUUDrksXAKX3ca1nW9wor2LX7+v1pKIjIzCIddFooS7m/hYaK9aSyIyYgqHXDfvNggW8TtTt/LCziP0qrUkIiOgcMh1oWK47Hau6XyDUx1dvLmvyeuKRCQLKBzyQSRKuLuZpQXv6hviRGREFA75YO4yCBbzUMUWXtx5hJ6YWksicm4Kh3wQKoL5d7Ck43U6urp5/d1jXlckIhlO4ZAvIlFC3Se4tWCvWksicl4Kh3wx9+MQKuGhii28tKuRrt641xWJSAZTOOSLYAHMv5OrO96gq7uL/9mr1pKIDC/gdQEyiSJRgtt/zG1Fe/k/r1Ry9GQXNVXF1FQWM6OsgIBfvyuISJLCIZ9ceguEp/BIxXY+eehKvv7znWdeCvqNWRVFzK4sYnZlMTWVRcxOBUd1RSFBBYdIXlE45JNUa2nh3hfY9Y1/4WhHgrqmdg40d1DXfPZxQ90J2rpjZzbz+4yZ5YXMriyiprL4zGNNVRHVFUUUBP0evikRSQeFQ76JRGHb09j+17hg3se5YEoB115S2W+Ic47m9h4ONLdT19SRfGxOPv58yyFOdp0NDjO4qCwVHFWpGUdlcsZx8dQiCkMKDpFspHDIN5feDOEyWPeP0HMKymdDRQ0UViT/pQfMjKqSMFUlYT48e+qgXbR09FDX3EFdU3u/GccLO45wvL2n39gLpxScnXFUnZ15zK4spiSsP34imUp/O/NNIAxXfwbe+ifY/9rZ9aFSqEgFRfns5PO+j6GiM0PLi0IsKgqxaFb5oN23dvbywZk21dkZx6/2HOXYxu5+Y6tKwn1mGqfPcSSXywqD6fovICIjYM7lxncLL1myxG3cuNHrMrJH10lo+QBO1EHLAThx4OzjiTqIdfYfXzy9f2D0DZEp1eA//+8Z7d0xDjSfbVP1nXkcOdnVb2xFUfDMlVQDz3WUFwWx1CxHRMbHzDY555YMWq9wkEGcg/ZjfQKjrn+ItNaD6/MhOvND2cyzLaqK2VBeczZMSqafaVkNp7MnzgfHB8846po6ONzaSd8/plMKAtRUFZ+dcfR5rCoJKThERkHhIBMnHoOTh4afdbQf7T8+WATlFw9oV9WcfV4w5ZyH647FOXi8s19o7E9dZVV/ooNEnz/CxSF/MiyqigaERzHTS8P4fAoOkb6GCwedc5DR8wdSraXZQ7/e05FsWfUNjNPPP3gLuk/2H19YMTgwTs8+ymcRDoSZO72EudNLBh8qluBQS2dyxtF0NjzeaTjFf+9sJNYnOXwGZYVBKopClBcFKU89VhSFqEgt931++jVdcSX5SDMHmVzOQeeJ4WcdrQch3veKJ4MpFw0+SX76nEfpDPAN/QG9WDxBQ2sXdc3t1DW1c/RUNyc6ejjR0UtrRy8nOnpoST129Ax/r6lwwHcmUCqKQlQUp8KjT9D0XV9RFKKsMIhfsxTJAmorSXZIJOBUw9CzjpYDcPIw0OfPrD8EZbOGudKqpt8luufSHYufCYoT7b20dPTQ0tknQNqTodLS0XNmXUtnL/HE8H9/phQEqCgOpQKjf5CUn5mp9F9fFPLrnIlMKrWVJDv4fMmT22UzYfb1g1+PdUPLQWipGzzrOPx2clbSV3jK2cAongbhEgiVJL8+NZR6Hi4hHCrmglAxF4RKoao4OS5YfM6rsJxznOqO0dKeCpU+M5Hk7CT5eKKjh+a2HvYdbaOlo7ffp88HCvl9qeAYGB5nn5cNaIWVFwV1exOZcAoHyS6BMFTNTf4MpevkgHZVXfJ58z44uB662wZfpnvO4xWcDZNwaSpUksFioRKmhEuYEirm4lBx8rMioWIoL4HpfQOo6uy2wSJ64o6Wzp5Ua+v07ORskJwOm5aOXvY3tbO5o4WWjh5648PPUkrDAcqLg2daWqfDozgcIBzwUxD0URD0Ew4kHwuCPsIBP+FB6/s8D/h0M8Y8pnCQ3FIwBS68MvkznEQcetqhpy352H0qtXx63en1bf2XT4/tOplsb50e390Gid4RFmiEQiVMDxUzPVTcZyZTcjZ4SkphanG/GY4LFdPlK+RkvIDWeJgTsSDNsRDHe4M0d1q/gGnp6OFAcwcnOnro7In3Oyk/Wn6fURDoHyDhM88HBMsQYRMO+AingmbYEEoF1en9hfw+XVWWARQOkn98/mSInOcS2lGJ9QwTLEMFTnvy1iVnAqcN2hr7B1ZPG7iz3/VtQGHq54JB7yd4dpYSLoFwMZSmAidYQMIXIu4LEbMgMV+QGEF6LUgvQXoI0mNBelyAboJ0JQJ0uQDdLkBnIkBnwk9nIkB73E9nwk9bPEB7LEF73EdbLMDxnjjdvQm6YnG6euN0xxJ09cbp6h3f95SHAj4KTgdL0Hc2eAL+AUHlGzQzCvgMM8MMfGb4DIyzy/3WD7VMatnXf7uz2w4xru9yatxw2/n6jB04zpdaHmqcz9f3GP1rLwkHCAUmdpancBCZCIEQBKZC0eB7UY2Jc9DbOXSQ9AwInu5hQqijGWJd+GI9+OLdBGPdySvBYl39gmdczAf+cPL9B8NQGAZ/CBcI4/xhEr4QCV+QuC9E3BckZqmQOh1OFqKHAL0E6XYBugnQnQjSSYDuRIBO56czHqAjkfqJ+2hv99Oe8NMS89OW+jkV83EqZnTGfSSjNL9874GrWX7FjAndp8JBJBOZJe9nFSoCpk38/uMxiHcnT/DHewY8didnQqcfY13Dv9ZvH11n1lmsG4v34Dv9Wm/LgH0MOLYb59fWGhAEF/IlZ1L+IPgC4AvgfAHwBXH+1KMFwB/A+YKp11LPzY87/Tw1LuE7vS6AsyAJXyC53vw4X5CEBUj4AsnH1PbJ5wHiqdccAeI+/5n1CQsQx39mTDy1bTy1PrnOT4IgMfPj8JFwDudIPpJ8TLjkRRHOwYIZEzgLTlE4iOQjf/IfSKyT1i0AAAO6SURBVELFXleSlIifI5jOFUh9HmPdWLwb4r2QiJ15tERvMgwTvWdfO/N66rV4e2o53mdcn+0Ssf77GG+YjYb5UkEXTP4/GxB++IMw/ZtQeeuEHlbhICLe8/lTs6Si8w7NCInE2ZAZFD6pkBkYMEMG08B1A1+L9x833GsFZRP+FhUOIiKj5fOBLwSEvK4kbXQRs4iIDKJwEBGRQRQOIiIyiMJBREQGUTiIiMggCgcRERlE4SAiIoMoHEREZJCc+SY4MzsGHBjj5lVA0wSWkw30nvOD3nN+GM97nu2cG3QDr5wJh/Ews41DfU1eLtN7zg96z/khHe9ZbSURERlE4SAiIoMoHJKe9LoAD+g95we95/ww4e9Z5xxERGQQzRxERGQQhYOIiAyS9+FgZsvNbI+Z7TOzx72uJ93M7CkzO2pmO7yuZbKY2Swze8XMdpnZTjP7ktc1pZOZFZhZrZltTb3fv/C6psliZn4ze9vM1nhdy2Qwszoz225mW8xs44TuO5/POZiZH9gLLAPqgQ3A/c65XZ4WlkZm9jGgDfg359wVXtczGcxsBjDDObfZzEqBTcDKXP3/bGYGFDvn2swsCLwBfMk592uPS0s7M/sjYAkwxTl3t9f1pJuZ1QFLnHMT/qG/fJ85fATY55x73znXAzwN3OtxTWnlnHsNOO51HZPJOdfgnNucen4K2A3M9Laq9HFJbanFYOon538LNLNq4C7gX7yuJRfkezjMBA72Wa4nh//REDCzGmAxsN7bStIr1V7ZAhwFXnLO5fT7TfkH4DEg4XUhk8gB/21mm8zsCxO543wPB8kjZlYC/BT4Q+fcSa/rSSfnXNw5twioBj5iZjndQjSzu4GjzrlNXtcyyT7qnLsauAP4/VTbeELkezgcAmb1Wa5OrZMck+q9/xT4D+fcf3ldz2RxzrUArwDLva4lzW4AVqR68E8Dt5jZv3tbUvo55w6lHo8Cz5BslU+IfA+HDcA8M5tjZiHgPuBZj2uSCZY6QfuvwG7n3N97XU+6mdk0MytPPS8kecHFO95WlV7Oua8656qdczUk/x7/yjn3gMdlpZWZFacusMDMioHbgAm7CjGvw8E5FwMeAV4keZLyx865nd5WlV5m9kPgLWC+mdWb2UNe1zQJbgA+Q/K3yS2pnzu9LiqNZgCvmNk2kr8AveScy4tLO/PMBcAbZrYVqAXWOudemKid5/WlrCIiMrS8njmIiMjQFA4iIjKIwkFERAZROIiIyCAKBxERGUThICIigygcRERkkP8PeB2a3Og+8WIAAAAASUVORK5CYII=\n",
            "text/plain": [
              "<Figure size 432x288 with 1 Axes>"
            ]
          },
          "metadata": {
            "tags": [],
            "needs_background": "light"
          }
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0F7x2ecwp_XY"
      },
      "source": [
        "# Customize the app introduction text\n",
        "You can change the text displayed in the landing page by changing the string below. It is markdown."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "sdGNx5Q6po9D"
      },
      "source": [
        "### This cell will write the string below to a .txt file, which will then be\n",
        "### parsed by the app as markdown and inserted at the top of the landing page\n",
        "\n",
        "introduction_text=\"\"\"\n",
        "### scdefg: scvi on Colab for differential expression with a Flask GUI\n",
        "\n",
        "*This app uses [scvi-tools](https://scvi-tools.org) and is made to run on a\n",
        " Google Colab instance with a GPU.   \n",
        " The code for this app and more information is available at\n",
        "[https://github.com/munfred/scdefg](https://github.com/munfred/scdefg)*\n",
        "\n",
        "By using a GPU scvi-tools allows you to quickly perform differential expression \n",
        "on single cell RNA sequencing data.   \n",
        "Just select cell types and experiments to compare and some genes to highlight.   \n",
        "It will produce an interactive volcano plot and a csv file with results in about 5 seconds! Try it! \n",
        "\"\"\"\n",
        "\n",
        "text_file = open(\"./scdefg/templates/introduction.txt\", \"w\")\n",
        "text_file.write(introduction_text)\n",
        "text_file.close()"
      ],
      "execution_count": 9,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "H7K4lSLgjcC9"
      },
      "source": [
        "# Run the cell below to start the app\n",
        "\n",
        "It will output a link to ngrok.io to access the app"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "aRG30muRgbdy",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "e65f14fc-d394-473c-a147-5426ea469692"
      },
      "source": [
        "logging.basicConfig(level=logging.INFO)\n",
        "logger=logging.getLogger(__name__)\n",
        "logger.info('Starting wormcells-de...')\n",
        "\n",
        "flask_app = Flask(__name__, template_folder='./scdefg/templates', static_folder='./scdefg/static')\n",
        "flask_app.config['JSONIFY_PRETTYPRINT_REGULAR'] = False\n",
        "flask_app.config['TEMPLATES_AUTO_RELOAD'] = True\n",
        "flask_app.config['SEND_FILE_MAX_AGE_DEFAULT'] = 0\n",
        "run_with_ngrok(flask_app)\n",
        "Misaka(flask_app, math_explicit = True)\n",
        "tables = Blueprint('tables', __name__, url_prefix='/tables')\n",
        "\n",
        "# read in the census dataframe with the number of cells of each label in each dataset\n",
        "# to render the table titles better we replace underscores with spaces,\n",
        "# and use non breaking hyphens (&#8209;) \n",
        "census =pd.read_csv(flask_app.open_resource('census.csv'))\n",
        "df_nice_names = census.copy()\n",
        "df_nice_names.columns = df_nice_names.columns.str.replace('_',' ')\n",
        "df_nice_names.columns = df_nice_names.columns.str.replace('-','&#8209;')\n",
        "\n",
        "# convert df to dict for sending as json to datatables\n",
        "dict_df = df_nice_names.to_dict(orient='records')\n",
        "# convert column names into dict for sending as json to datatables\n",
        "columns = [{\"data\": item, \"title\": item} for item in df_nice_names.columns]\n",
        "\n",
        "#### datatables to render the selection tables ####\n",
        "@tables.route(\"/\", methods=['GET'])\n",
        "def clientside_table_content():\n",
        "    return jsonify({'data': dict_df, 'columns': columns})\n",
        "\n",
        "flask_app.register_blueprint(tables)\n",
        "\n",
        "# this is the landing page\n",
        "@flask_app.route(\"/\")\n",
        "def clientside_table():\n",
        "    return render_template(\"clientside_table.html\")\n",
        "\n",
        "@flask_app.route(\"/test\")\n",
        "def test():\n",
        "    return render_template(\"test.html\")\n",
        "\n",
        "@flask_app.route('/submit', methods=['POST', 'GET'])\n",
        "def receive_submission():\n",
        "    logger.info('Got a submission!')\n",
        "    timestamp = time.strftime(\"%Y-%m-%d_%H:%M:%S\")\n",
        "\n",
        "    # answer is a dict of json strings containing selected row and column index numbers\n",
        "    answer = request.form.to_dict(flat=False)   \n",
        "    # need to convert the json strings to dict, then to a data frame\n",
        "    # data1 is the selection for the first group, data2 for the second\n",
        "    data1 = json.loads(answer['data1'][0])\n",
        "    data1_df = pd.DataFrame.from_dict(data1[0])\n",
        "    data2 = json.loads(answer['data2'][0])\n",
        "    data2_df = pd.DataFrame.from_dict(data2[0])\n",
        "\n",
        "    # now map the index number to experiment name and cell type name\n",
        "    group1 = pd.DataFrame()\n",
        "    group1['cell_type1'] = data1_df['row'].map(census['Cell Type'])\n",
        "    group1['experiment1'] = data1_df['column'].map(pd.Series(census.columns.values))\n",
        "\n",
        "    group2 = pd.DataFrame()\n",
        "    group2['cell_type2'] = data2_df['row'].map(census['Cell Type'])\n",
        "    group2['experiment2'] = data2_df['column'].map(pd.Series(census.columns.values))\n",
        "\n",
        "    genes = StringIO(json.loads(answer['genes'][0]))\n",
        "    genes_df = pd.read_csv(genes, names=['selected_genes'])\n",
        "\n",
        "    selected_groups_df = pd.concat([group1, group2, genes_df], axis=1)\n",
        "    selected_groups_df.to_csv('selected_groups.csv')\n",
        "    selected_groups_df.to_csv(timestamp+'selected_groups.csv')\n",
        "\n",
        "#### Creates the masks for the selected cell groups\n",
        "\n",
        "    # first create the mask as an array of all false\n",
        "    # then for each group in the data add them to the mask\n",
        "    group1_mask = adata.obs['experiment_code'] != adata.obs['experiment_code']\n",
        "    for idx, row in group1.iterrows():\n",
        "        mask = (adata.obs['experiment_code']==row['experiment1']) & (adata.obs['cell_type']==row['cell_type1'])\n",
        "        group1_mask = group1_mask | mask\n",
        "\n",
        "    group2_mask = adata.obs['experiment_code'] != adata.obs['experiment_code']\n",
        "    for idx, row in group2.iterrows():\n",
        "        mask = (adata.obs['experiment_code']==row['experiment2']) & (adata.obs['cell_type']==row['cell_type2'])\n",
        "        group2_mask = group2_mask | mask\n",
        "\n",
        "    # the masks then define the two groups of cells on which to perform DE\n",
        "    fdr_target=0.005\n",
        "    de = model.differential_expression( adata, \n",
        "                                       idx1=group1_mask, \n",
        "                                       idx2=group2_mask,\n",
        "                                       fdr_target=fdr_target)\n",
        "\n",
        "#### Wrangles the DE results dataframe a bit\n",
        "\n",
        "    # first we create these variables to customize the hover text in plotly's heatmap\n",
        "    # the text needs to be arranged in a matrix the same shape as the heatmap\n",
        "    #try to add gene descriptions and gene names if the adata has those, otherwise add a blank\n",
        "    try:\n",
        "        de['gene_description']=de.index.map(adata.var['gene_description'])\n",
        "        # for the gene descriptions text, which can be long, we add line breaks\n",
        "        de['gene_description_html'] = de['gene_description'].str.wrap(80).str.replace('\\n','<br>')\n",
        "        de['gene_name']=de.index.map(adata.var['gene_name'])\n",
        "        de['gene_id'] = de.index.map(adata.var['gene_id'])\n",
        "\n",
        "    except:\n",
        "        de['gene_description_html'] = 'gene description here'\n",
        "        de['gene_name']='gene name here'\n",
        "        de['gene_id']='gene ID here'\n",
        "\n",
        "    de['gene_name']=de['gene_name'].fillna('-')\n",
        "    # calculate the -log10(p-value) for the volcano\n",
        "    de['minuslog10pval']=-np.log10(de['proba_not_de'] + 0.00001)\n",
        "\n",
        "    # all genes are initially colored black\n",
        "    de['color'] = 'black'\n",
        "    # uncomment line below to color genes by FDR significance\n",
        "    # de['color'] = de['is_de_fdr_'+str(fdr_target)].map({True:'steelblue',False:'gray'})\n",
        "\n",
        "    # then we loops through the list of genes provided to color some red\n",
        "    # gene ids should be a perfect match\n",
        "    pd.set_option('mode.chained_assignment',None) #supress warning\n",
        "    de['color'][de['gene_id'].isin(genes_df['selected_genes'].values)] = 'red'\n",
        "    # gene names should be a partial match\n",
        "    for partial_string in genes_df['selected_genes'].values:\n",
        "        de['color'][de['gene_name'].str.contains(partial_string)] = 'red'\n",
        "\n",
        "#### This makes the volcano plot using plotly\n",
        "    fig = go.Figure(\n",
        "                    data=go.Scatter(\n",
        "                            x=de[\"lfc_mean\"].round(3)\n",
        "                            , y=de[\"minuslog10pval\"].round(3)\n",
        "                            , mode='markers'\n",
        "                            , marker=dict(\n",
        "                                        color=de['color'],\n",
        "                                        opacity=0.5)\n",
        "                            , hoverinfo='text'\n",
        "                            , text=de['gene_description_html']\n",
        "                            , customdata=de.gene_name + '<br>' + de.gene_id\n",
        "                            , hovertemplate='%{customdata} <br>' +\n",
        "                                            '-log10 p-value: %{y}<br>' +\n",
        "                                            'Mean log fold change: %{x}' +\n",
        "                                            '<extra>%{text}</extra>'\n",
        "                            )\n",
        "                            , layout= {\n",
        "                                    \"title\": {\"text\": \n",
        "                                            \"Differential expression on group 1 and group 2 <br> Dashes mark p = 0.01\" \n",
        "                                            , 'x':0.5        \n",
        "                                            }\n",
        "                                    , 'xaxis': {'title': {\"text\": \"Mean log fold change\"}}\n",
        "                                    , 'yaxis': {'title': {\"text\": \"-log10 p-value\"}}\n",
        "        #                             , \"height\": 600,\n",
        "        #                             , \"width\":1000\n",
        "                            }\n",
        "                )\n",
        "    fig.update_layout(hovermode='closest', template='none')\n",
        "    fig.add_shape(type=\"line\", x0=-6, y0=2, x1=6, y1=2, line=dict(color=\"lightsalmon\", width=2, dash=\"dash\"))\n",
        "\n",
        "    # overwrites the last figure in order to serve it in the results page\n",
        "    fig.write_html('./scdefg/static/fig.html')\n",
        "    de_csv = de[['gene_name','minuslog10pval','lfc_mean','lfc_std','proba_not_de','is_de_fdr_'+str(fdr_target)]].round(2)\n",
        "    de_csv.to_csv('results.csv')\n",
        "    de_csv.to_csv(timestamp+'results.csv')\n",
        "    # also writes the figure in the colab folder with a timestamp\n",
        "    fig.write_html( timestamp + '_fig.html')\n",
        "\n",
        "\n",
        "    return redirect(url_for('test'))\n",
        "\n",
        "@flask_app.route(\"/get_results_csv\")\n",
        "def get_results_csv():\n",
        "    with open(\"results.csv\") as fp:\n",
        "        csv = fp.read()  \n",
        "    return Response( csv, mimetype=\"text/csv\",\n",
        "        headers={\"Content-disposition\":\"attachment; filename=DE_results.csv\"})\n",
        "\n",
        "@flask_app.route(\"/get_groups_csv\")\n",
        "def get_groups_csv():\n",
        "    with open(\"selected_groups.csv\") as fp:\n",
        "        csv = fp.read()  \n",
        "    return Response( csv, mimetype=\"text/csv\",\n",
        "        headers={\"Content-disposition\":\"attachment; filename=selected_groups.csv\"})\n",
        "\n",
        "@flask_app.route(\"/get_plot\")\n",
        "def get_plot():\n",
        "    with open(\"./scdefg/static/fig.html\") as fp:\n",
        "        csv = fp.read()  \n",
        "    return Response( csv, mimetype=\"text/html\",\n",
        "        headers={\"Content-disposition\":\"attachment; filename=fig.html\"})\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "    flask_app.run()"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "INFO:werkzeug: * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "stream",
          "text": [
            " * Running on http://00f1200167a6.ngrok.io\n",
            " * Traffic stats available on http://127.0.0.1:4040\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "stream",
          "text": [
            "INFO:werkzeug:127.0.0.1 - - [26/Jan/2021 17:24:56] \"\u001b[37mGET / HTTP/1.1\u001b[0m\" 200 -\n",
            "INFO:werkzeug:127.0.0.1 - - [26/Jan/2021 17:24:56] \"\u001b[37mGET /static/js/clientside_table.js HTTP/1.1\u001b[0m\" 200 -\n",
            "INFO:werkzeug:127.0.0.1 - - [26/Jan/2021 17:24:57] \"\u001b[37mGET /tables/ HTTP/1.1\u001b[0m\" 200 -\n",
            "INFO:werkzeug:127.0.0.1 - - [26/Jan/2021 17:24:57] \"\u001b[37mGET /static/favicon.ico HTTP/1.1\u001b[0m\" 200 -\n",
            "INFO:werkzeug:127.0.0.1 - - [26/Jan/2021 17:24:57] \"\u001b[37mGET /tables/ HTTP/1.1\u001b[0m\" 200 -\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "stream",
          "text": [
            "DE...: 100%|██████████| 1/1 [00:05<00:00,  5.67s/it]\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "stream",
          "text": [
            "INFO:werkzeug:127.0.0.1 - - [26/Jan/2021 17:25:15] \"\u001b[32mPOST /submit HTTP/1.1\u001b[0m\" 302 -\n",
            "INFO:werkzeug:127.0.0.1 - - [26/Jan/2021 17:25:15] \"\u001b[37mGET /test HTTP/1.1\u001b[0m\" 200 -\n",
            "INFO:werkzeug:127.0.0.1 - - [26/Jan/2021 17:25:15] \"\u001b[37mGET /test HTTP/1.1\u001b[0m\" 200 -\n",
            "INFO:werkzeug:127.0.0.1 - - [26/Jan/2021 17:25:15] \"\u001b[37mGET /static/fig.html HTTP/1.1\u001b[0m\" 200 -\n",
            "INFO:werkzeug:127.0.0.1 - - [26/Jan/2021 17:25:19] \"\u001b[33mGET /favicon.ico HTTP/1.1\u001b[0m\" 404 -\n"
          ],
          "name": "stderr"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "aBvAsfw3pYZS"
      },
      "source": [
        ""
      ],
      "execution_count": 14,
      "outputs": []
    }
  ]
}